x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Client, errors } from '@elastic/elasticsearch';
+import { Logger } from '@kbn/logging';
+
+const INDEX_NAME = '.kibana-distributed-lock-manager';
+
+export enum LockId {
+  KnowledgeBaseReindex = 'knowledge_base_reindex',
+}
+
+export interface LockDocument {
+  createdAt: string;
+  expiresAt: string;
+  meta: Record<string, any>;
+}
+
+export class LockManager {
+  constructor(private lockId: LockId, private esClient: Client, private logger: Logger) {}
+
+  /**
+   * Attempts to acquire a lock by creating a document with the given lockId.
+   * If the lock exists and is expired, it will be released and acquisition retried.
+   */
+  public async acquire(
```

Comment:
wdyt of having this take a callback that returns a promise, and managing locks internally?

Reply:
WDYT about `withLock`?

Reply:
sgtm!